src/kbmod/search/kernels.cu

```diff
@@ -19,11 +19,62 @@
 
 #include "common.h"
 #include "cuda_errors.h"
-
-//#include "image_stack.h"
+#include "psi_phi_array_ds.h"
 
 namespace search {
 
+extern "C" void device_allocate_psi_phi_array(PsiPhiArray* data) {
+    if (!data->cpu_array_allocated())
+        throw std::runtime_error("CPU data is not allocated.");
+    if (data->gpu_array_allocated())
+        throw std::runtime_error("GPU data is already allocated.");
+
+    void* device_array_ptr;
+    checkCudaErrors(cudaMalloc((void **)&device_array_ptr, data->get_total_array_size()));
+    checkCudaErrors(cudaMemcpy(device_array_ptr,
+                               data->get_cpu_array_ptr(),
+                               data->get_total_array_size(),
+                               cudaMemcpyHostToDevice));
+    data->set_gpu_array_ptr(device_array_ptr);
+}
+
+extern "C" void device_free_psi_phi_array(PsiPhiArray* data) {
+    if (data->gpu_array_allocated()) {
+        checkCudaErrors(cudaFree(data->get_gpu_array_ptr()));
+        data->set_gpu_array_ptr(nullptr);
+    }
+}
+
+__forceinline__ __device__ PsiPhi read_encoded_psi_phi(PsiPhiArrayMeta &params, void *psi_phi_vect, int time,
```

Comment:
Yes. I broke this out for two reasons:
1) It makes the code a little easier to read.
2) It keeps all the decoding logic in a single place in case we want to change it later.
The `__forceinline__` inlines this code in the function, so we don't pay the overhead of a function call.